### PR TITLE
Support cooperating responders

### PR DESCRIPTION
**Pull request by mattsaxon**
Initial commit to address
https://github.com/jstasiak/python-zeroconf/issues/163

### CODE REVIEW

File: zeroconf/__init__.py [hunk 1]
```diff
@@ -2197,7 +2197,11 @@ def remove_all_service_listeners(self) -> None:
             self.remove_service_listener(listener)
 
     def register_service(
-        self, info: ServiceInfo, ttl: Optional[int] = None, allow_name_change: bool = False
+        self,
+        info: ServiceInfo,
+        ttl: Optional[int] = None,
+        allow_name_change: bool = False,
+        cooperating_responders: bool = False,
```
Comment (jstasiak) [inline zeroconf/__init__.py]: Can you document the use case for the new parameter maybe? It'll be helpful to have a few words in the [Sphinx-built documentation](https://python-zeroconf.readthedocs.io/en/latest/api.html#zeroconf.Zeroconf.register_service).

Comment (mattsaxon) [reply zeroconf/__init__.py]: Updated PR with a commit as requested

Comment (jstasiak) [reply zeroconf/__init__.py]: Cool, I allowed myself to add few more words there.

**Comment by mattsaxon** [comment]
No, I don't have merged rights on this repo.

**Comment by jstasiak** [comment]
Well, I'd swear I gave you access, my memory is playing tricks on me apparently. Check your inbox.

**Comment by jstasiak** [comment]
Oops, I just now realized the way I phrased it was ambiguous. I meant have a look at your inbox *now*, as I sent you an invitation. :)

**Comment by mattsaxon** [comment]
Don’t worry, it was clear to me, especially when the invite came through at the same time ;-)

File: zeroconf/__init__.py [hunk 2]
```diff
@@ -2339,7 +2345,7 @@ def unregister_all_services(self) -> None:
                 i += 1
                 next_time += _UNREGISTER_TIME
 
-    def check_service(self, info: ServiceInfo, allow_name_change: bool) -> None:
+    def check_service(self, info: ServiceInfo, allow_name_change: bool, cooperating_responders: bool) -> None:
```
Comment (dtantsur) [inline zeroconf/__init__.py]: This is essentially a breaking change, why not default to `False`?

File: zeroconf/__init__.py [hunk 3]
```diff
@@ -2197,18 +2197,24 @@ def remove_all_service_listeners(self) -> None:
...
+        allow_name_change: bool = False,
+        cooperating_responders: bool = False,
     ) -> None:
         """Registers service information to the network with a default TTL.
         Zeroconf will then respond to requests for information for that
         service.  The name of the service may be changed if needed to make
-        it unique on the network."""
+        it unique on the network. Additionally multiple cooperating responders
+        can register the same service on the network for resilience
+        (if you want this behavior set `cooperating_responders` to `True`)."""
```
Comment (dtantsur) [inline zeroconf/__init__.py]: Should we state that we're not exactly RFC compliant without it (and, well, with it probably as well)?

File: zeroconf/__init__.py (hunk 2 above)
Comment (jstasiak) [reply zeroconf/__init__.py]: Only if `check_service()` is a public API I think? (I'm not entirely sure if it should be or not)

File: zeroconf/__init__.py (hunk 3 above)
Comment (jstasiak) [reply zeroconf/__init__.py]: Hmm, possibly. Also, would always enabling this break things? Maybe we don't need that switch at all?

File: zeroconf/__init__.py (hunk 2 above)
Comment (dtantsur) [reply zeroconf/__init__.py]: I assume that everything is public unless it starts with an underscore :) But yes, I don't remember whether this function is intended to be public.

File: zeroconf/__init__.py (hunk 3 above)
Comment (dtantsur) [reply zeroconf/__init__.py]: If this is not intended, you may end up with conflicting responders on the network.. IIRC (I've pasted the quote somewhere) RFC allows several responders announcing the same information. I wonder if we should distinguish between the cases of the same information and different information.

Comment (mattsaxon) [reply zeroconf/__init__.py]: I agree that this is needed.

Is easy to make it backwards compatible as you say.

So we should either do that or change it to private.

From my perspective, this is a private method, I can’t see any reason to call it direct as it essentially registers the name on the network, so it only makes sense as part of calling register_service()

Comment (mattsaxon) [reply zeroconf/__init__.py]: I've updated so it is backwards compatible and we can have the debate about private methods in another PR!

Comment (jstasiak) [reply zeroconf/__init__.py]: Yeah, sounds good

**Comment by jstasiak** [comment]
This has just been released in 0.24.5.

**Pull request merged**